### ISSUE TITLE
Foldable instance for IntMap is broken? (internally inconsistent, unlawful)

### ISSUE THREAD

**Issue by mstksg** [haskell/containers#579]
Was running some property tests and noticed that the order of traversals for `Foldable` methods for `IntMap` is internally inconsistent.  `foldr`, `foldl`, `toList` appear to traverse items in order of ascending keys (which I believe is the intended behavior, for consistency with `Map k`'s instance)

```haskell
m = IM.fromList $ zip [-5..5] ['a'..]

foldr (:) [] m
-- "abcdefghijk"

traverse_ (Const . (:[])) m
-- Const "abcdefghijk"

toList m
-- "abcdefghijk"
```

However, `foldMap` and `fold` (and `foldMapWithKey`) and `traverse` (and `traverseWithKey`) appear to traverse items at positive keys first, then items at negative keys second:

```haskell
foldMap (:[]) m
-- "fghijkabcde"

traverse (Const . (:[])) m
-- Const "fghijkabcde"
```

(Note that `traverse (Const . (:[]))` is different than `traverse_ (Const . (:[])`)

The reason appears to be that an in-order traversal for `IntMap` is a positive-then-negative traversal (due to how its internal structure is based on the bitstring represenation of Int, instead of the Ord instance like for Map).  `foldr` and `foldl` explicitly correct for this (and so also the default implementation of `traverse_`), but `fold` and `foldMap` and `traverse` do not (they are implemented the same way as they are for `Map`).

The consensus in #535 seems to be that the order of traversals for maps isn't particularly significant.  However, the traversal order should at least be internally consistent within different `Foldable` methods, because of the Foldable "laws":

```haskell
foldr f z t = appEndo (foldMap (Endo . f) t ) z
foldl f z t = appEndo (getDual (foldMap (Dual . Endo . flip f) t)) z
```

This implies that:

```haskell
foldMap (:[]) == foldr (:) []
```

which I believe is a reasonable expectation that is currently being broken.

In order to preserve the lawfulness/expectation of consistency of the `Foldable` instance, I think one of the three options are possible:

1. Change `foldMap` and `fold` and `traverse` to match the behavior of `foldr` and `foldl` and `toList`, and traverse items in order of ascending key by correcting for the internal structure of the map.  Presumably, `foldMapWithKey` would also need to be changed as well, to traverse in order of ascending keys.  This way makes the most sense to me (making the instances consistent with `Map k`), but is a potentially breaking change to anyone using `foldMap` and `fold` with noncommutative monoids (and `traverse` with noncommutative applicatives).
2. Change `Foldable.foldr` and `Foldable.foldl` and `Foldable.toList` to match the behavior of `foldMap` and `fold` (traverse positive, then negative), by making them *different* from `Data.IntMap.foldr` and `Data.IntMap.foldl` (traverse negative, then positive).  This keeps `Data.IntMap.foldr` and `Data.IntMap.foldl` consistent with `Data.Map.foldr` and `Data.Map.foldl`, and also fixes the inconsistencies in the `Foldable` instance.  (This method would also break `Foldable.toList == elems`, but I don't think this is a particularly important thing)
3. Change `Data.IntMap.foldr` and `Data.IntMap.foldl` to match the behavior of `foldMap` and `fold` (traverse positive, then negative), and so also `Foldable.foldr` and `Foldable.foldl` to match.  This fixes the inconsistencies in the `Foldable` instances, but breaks consistency with `Data.Map.foldr` and `Data.Map.foldl`.  This method also breaks `Foldable.toList == elems`.

To summarize:

| | Keep Current Situation | Change `foldMap`, `traverse` | Change `Foldable.foldr` | Change `Foldable.foldr` and `IntMap.foldr` |
|-|-|-|-|-|
| Internally consistent | | x | x | x |
| `Foldable.foldr` = `IntMap.foldr` | x | x | | x |
| Consistent with `Map` | | x | | |
| `Foldable.toList` = `elems` | x | x | | |
| Doesn't break code using non-commutative monoids or applicatives | x | | x | x |

Note that the `Traversable` instances are internally consistent at the moment (traversing positives, the negatives), but also are technically "unlawful" because they are not consistent with `Foldable` methods.  (2) and (3) would fix this, but (1) would require the `Traversable` instance to also change.

**Comment by treeowl** [comment]
Ugh! That's awful! I definitely agree that option (1) is the correct one. @wrengr, do you have the time to look at this?
